gcc/config/riscv/riscv-insert-vsetvli.cc

```diff
@@ -0,0 +1,1853 @@
+	    // against was created in this block. We need the first and third
+	    // phase to treat the store the same way.
+	    if (!canskipvsetvliforloadstore (insn, new_info, info.change)
+		&& needvsetvli (insn, new_info, info.change))
+	      info.change = new_info;
+	  }
+      }
+    // If this is something that updates VL/VTYPE that we don't know about, set
+    // the state to unknown.
+    if (update_vlvtyp_p (insn))
+      info.change = vinfo::getunknown ();
+  }
+
+  // Initial exit state is whatever change we found in the block.
+  info.exit = info.change;
+
+  return vector_p;
+}
+
+static void
+computeincomingvlvtype (const basic_block bb)
+{
+  bb_vinfo &info = m_bb_vinfo_map[bb->index];
+  info.inqueue = false;
+
+  vinfo in_info;
+  if (EDGE_COUNT (bb->preds) == 0)
+    {
+      in_info.setunknown ();
+    }
+  else
+    {
+      edge e;
+      edge_iterator ei;
+      FOR_EACH_EDGE (e, ei, bb->preds)
+      {
+	basic_block ancestor = e->src;
+	in_info = in_info.intersect (m_bb_vinfo_map[ancestor->index].exit);
+      }
+    }
+
+  // If we don't have any valid predecessor value, wait until we do.
+  if (!in_info.valid_p ())
+    return;
+
+  info.pred = in_info;
+
+  vinfo merge_info = info.pred.merge (info.change);
+
+  // If the new exit value matches the old exit value, we don't need to revisit
+  // any blocks.
+  if (info.exit == merge_info)
+    return;
+
+  info.exit = merge_info;
+
+  // Add the successors to the work list so we can propagate the changed exit
+  // status.
+  edge e;
+  edge_iterator ei;
+  FOR_EACH_EDGE (e, ei, bb->succs)
+  {
+    basic_block succ = e->dest;
+    if (!m_bb_vinfo_map[succ->index].inqueue)
+      m_bb_queue.push (succ);
+  }
+}
+
+static void
+insertvsetvli (rtx_insn *insn, const vinfo &curr_info, const vinfo &prev_info)
+{
+  extract_insn_cached (insn);
+  rtx avl = curr_info.getavl ();
+  rtx vtype = GEN_INT (curr_info.encodevtype ());
+  rtx zero = gen_rtx_REG (Pmode, X0_REGNUM);
+
+  // Use X0, X0 form if the AVL is the same and the SEW+LMUL gives the same
+  // VLMAX
+  if (prev_info.valid_p () && !prev_info.unknown_p ()
+      && curr_info.avl_equal_p (prev_info)
+      && curr_info.vlmax_equal_p (prev_info))
+    {
+      emit_vsetvl_insn (zero, zero, vtype, insn);
+      return;
+    }
+
+  if (curr_info.getavl () == NULL_RTX)
+    {
+      if (prev_info.valid_p () && !prev_info.unknown_p ()
+	  && curr_info.vlmax_equal_p (prev_info))
+	{
+	  emit_vsetvl_insn (zero, zero, vtype, insn);
+	  return;
+	}
+      // Otherwise use an AVL of 0 to avoid depending on previous vl.
+      emit_vsetvl_insn (zero, GEN_INT (0), vtype, insn);
+      return;
+    }
+
+  if (rtx_equal_p (curr_info.getavl (), gen_rtx_REG (Pmode, X0_REGNUM)))
+    {
+      avl = gen_reg_rtx (Pmode);
+      emit_vsetvl_insn (avl, gen_rtx_REG (Pmode, X0_REGNUM), vtype, insn);
+      return;
+    }
+
+  emit_vsetvl_insn (zero, avl, vtype, insn);
+}
+
+static void
+emitvsetvlis (const basic_block bb)
```

Comment:
emit_vsetvlis

Reply:
> emit_vsetvlis

OK. Thanks for your suggestions. I am going to collect all your suggestions on revision and modify them into a single commit. So that I won't missed any information you gave.

Reply:
> emit_vsetvlis

Hi, kito. I fixed insert vsetvlic PASS issue: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1bc0aa3c0319eb447468cdd074df80b4bc46267a. Thanks~